src/model/chathistory.cpp

```diff
@@ -0,0 +1,456 @@
+        }
+
+        auto username = coreIdHandler.getUsername();
+
+        auto onInsertion = [this, id](RowId historyId) { handleDispatchedMessage(id, historyId); };
+
+        history->addNewMessage(friendPk, content, selfPk, message.timestamp, false, username,
+                               onInsertion);
+    }
+
+    sessionChatLog.onMessageSent(id, message);
+}
+
+void ChatHistory::onMessageComplete(DispatchedMessageId id)
+{
+    if (canUseHistory()) {
+        completeMessage(id);
+    }
+
+    sessionChatLog.onMessageComplete(id);
+}
+
+/**
+ * @brief Forces the given index and all future indexes to be in the chatlog
+ * @param[in] idx
+ * @note Marked const since this doesn't change _external_ state of the class. We
+     still have all the same items at all the same indexes, we've just stuckem
+     in ram
+ */
+void ChatHistory::ensureIdxInSessionChatLog(ChatLogIdx idx) const
+{
+    if (needsLoadFromHistory(idx, sessionChatLog)) {
+        loadHistoryIntoSessionChatLog(idx);
+    }
+}
+/**
+ * @brief Unconditionally loads the given index and all future messages that
+ * are not in the session chat log into the session chat log
+ * @param[in] idx
+ * @note Marked const since this doesn't change _external_ state of the class. We
+   still have all the same items at all the same indexes, we've just stuckem
+   in ram
+ * @note no end idx as we always load from start -> latest. In the future we
+ * could have a less contiguous history
+ */
+void ChatHistory::loadHistoryIntoSessionChatLog(ChatLogIdx start) const
+{
+    if (!needsLoadFromHistory(start, sessionChatLog)) {
+        return;
+    }
+
+    auto end = sessionChatLog.getFirstIdx();
+
+    // We know that both history and us have a start index of 0 so the type
+    // conversion should be safe
+    assert(getFirstIdx() == ChatLogIdx(0));
+    auto messages = history->getMessagesForFriend(f.getPublicKey(), start.get(), end.get());
+
+    assert(messages.size() == end.get() - start.get());
+    ChatLogIdx nextIdx = start;
+
+    for (const auto& message : messages) {
+        // Note that message.id is _not_ a valid conversion here since it is a
+        // global id not a per-chat id like the ChatLogIdx
+        auto currentIdx = nextIdx++;
+        auto sender = ToxId(message.sender).getPublicKey();
+        switch (message.content.getType()) {
+        case HistMessageContentType::file: {
+            const auto date = message.timestamp;
+            const auto file = message.content.asFile();
+            const auto chatLogFile = ChatLogFile{date, file};
+            sessionChatLog.insertFileAtIdx(currentIdx, sender, message.dispName, chatLogFile);
+            break;
+        }
+        case HistMessageContentType::message: {
+            auto messageContent = message.content.asMessage();
+
+            auto isAction = handleActionPrefix(messageContent);
+
+            // It's okay to skip the message processor here because we can
```

Comment:
I think you're right the phrasing here doesn't really hit the root of the problem. It's not really a precondition thing on second thought. I think it's more of a message processor being the conversion between toxcore format and internal representation. I'll update the comment to reflect that since we aren't sending the message we don't have to process it.

Reply:
Done, please review new comment :)